File: web/src/lib/editor/keymap.ts
import type { EditorState } from './state';
import { undo, redo } from './state';
import * as M from './motions';
import * as E from './edits';
import * as I from './insert';
import { runCommand, type AppEffect, type CommandCtx } from './commands';

export interface KeyInput {
  key: string;
  ctrl: boolean;
  meta: boolean;
  shift: boolean;
}
export interface KeyResult {
  state: EditorState;
  effect?: AppEffect;
}

export function handleKey(state: EditorState, key: KeyInput, ctx: CommandCtx): KeyResult {
  if (state.command !== null) return handleCommandMode(state, key, ctx);
  if (state.mode === 'insert') return handleInsertMode(state, key);
  return handleNormalMode(state, key, ctx);
}

function handleCommandMode(state: EditorState, key: KeyInput, ctx: CommandCtx): KeyResult {
  if (key.key === 'Escape') return { state: { ...state, command: null, message: '' } };
  if (key.key === 'Enter') return runCommand(state, ctx);
  if (key.key === 'Backspace') return { state: { ...state, command: (state.command ?? '').slice(0, -1) } };
  if (key.key.length === 1 && !key.ctrl && !key.meta) {
    return { state: { ...state, command: (state.command ?? '') + key.key } };
  }
  return { state };
}

function handleInsertMode(state: EditorState, key: KeyInput): KeyResult {
  switch (key.key) {
    case 'Escape':
      return { state: I.exitInsert(state) };
    case 'Backspace':
      return { state: I.backspace(state) };
    case 'Enter':
      return { state: I.insertNewline(state) };
    case 'Tab':
      return { state: I.insertTab(state) };
    case 'ArrowLeft':
      return { state: M.moveLeft(state) };
    case 'ArrowRight':
      return { state: M.moveRight(state) };
    case 'ArrowUp':
      return { state: M.moveUp(state) };
    case 'ArrowDown':
      return { state: M.moveDown(state) };
  }
  if (key.ctrl && (key.key === 'w' || key.key === 'W')) return { state: I.deleteWordBefore(state) };
  if (key.key.length === 1 && !key.ctrl && !key.meta) return { state: I.insertText(state, key.key) };
  return { state };
}

function handleNormalMode(state: EditorState, key: KeyInput, ctx: CommandCtx): KeyResult {
  if (state.pending === 'g') {
    const s = { ...state, pending: '' as const };
    if (key.key === 'g') return { state: M.firstLine(s) };
    if (key.key === 't') return { state: s, effect: { type: 'tabNext' } };
    if (key.key === 'T') return { state: s, effect: { type: 'tabPrev' } };
    return handleNormalMode(s, key, ctx);
  }
  if (state.pending === 'd') {
    const s = { ...state, pending: '' as const };
    if (key.key === 'd') return { state: E.deleteLine(s) };
    return handleNormalMode(s, key, ctx);
  }
  if (state.pending === 'y') {
    const s = { ...state, pending: '' as const };
    if (key.key === 'y') return { state: E.yankLine(s) };
    return handleNormalMode(s, key, ctx);
  }

  switch (key.key) {
    case 'h':
    case 'ArrowLeft':
      return { state: M.moveLeft(state) };
    case 'l':
    case 'ArrowRight':
      return { state: M.moveRight(state) };
    case 'j':
    case 'ArrowDown':
      return { state: M.moveDown(state) };
    case 'k':
    case 'ArrowUp':
      return { state: M.moveUp(state) };
    case 'w':
      return { state: M.wordForward(state) };
    case 'b':
      return { state: M.wordBackward(state) };
    case 'e':
      return { state: M.wordEnd(state) };
    case '0':
      return { state: M.lineStart(state) };
    case '$':
      return { state: M.lineEnd(state) };
    case 'G':
      return { state: M.lastLine(state) };
    case 'g':
      return { state: { ...state, pending: 'g' } };
    case 'd':
      return { state: { ...state, pending: 'd' } };
    case 'y':
      return { state: { ...state, pending: 'y' } };
    case 'x':
      return { state: E.deleteChar(state) };
    case 'p':
      return { state: E.pasteBelow(state) };
    case 'P':
      return { state: E.pasteAbove(state) };
    case 't':
      return { state: E.toggleTodo(state) };
    case 'u':
      return { state: undo(state) };
    case 'i':
      return { state: I.enterInsert(state) };
    case 'a':
      return { state: I.enterInsertAfter(state) };
    case 'A':
      return { state: I.enterInsertLineEnd(state) };
    case 'o':
      return { state: I.openBelow(state) };
    case 'O':
      return { state: I.openAbove(state) };
    case ':':
      return { state: { ...state, command: '', message: '' } };
    case 'Enter':
      return { state: M.moveDown(state) };
    case '[':
      return { state, effect: { type: 'prevDay' } };
    case ']':
      return { state, effect: { type: 'nextDay' } };
    case 'Escape':
      return { state: I.enterInsert({ ...state, pending: '', message: '' }) };
  }
  if (key.ctrl && (key.key === 'r' || key.key === 'R')) return { state: redo(state) };
  if (key.ctrl && (key.key === 't' || key.key === 'T')) return { state, effect: { type: 'today' } };
  return { state };
}

File: crates/slugline-core/src/editor/keymap.rs
use super::state::{EditorState, Mode, Pending};
use super::{edits, insert, motions, state};

/// A normalized key event. `key` follows the DOM `KeyboardEvent.key` convention
/// ("h", "A", "Enter", "Escape", "ArrowLeft", " ", ...).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct KeyInput {
    pub key: String,
    pub ctrl: bool,
    pub meta: bool,
    pub shift: bool,
}

/// Side effects the app performs (navigation, save, theme). Handlers land in Phase 2/5;
/// the variants exist now so signatures are stable.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum AppEffect {
    Goto(String),
    Today,
    Tab(String),
    Close,
    Save,
    PrevDay,
    NextDay,
    TabNext,
    TabPrev,
    Theme(String),
}

pub struct KeyResult {
    pub state: EditorState,
    pub effect: Option<AppEffect>,
}

fn state_only(state: EditorState) -> KeyResult {
    KeyResult {
        state,
        effect: None,
    }
}

fn state_effect(state: EditorState, effect: AppEffect) -> KeyResult {
    KeyResult {
        state,
        effect: Some(effect),
    }
}

pub fn handle_key(state: &EditorState, key: &KeyInput) -> KeyResult {
    // Phase 1: command mode (`:`) is never entered; it is added in Phase 5.
    if state.mode == Mode::Insert {
        return state_only(handle_insert(state, key));
    }
    handle_normal(state, key)
}

fn handle_insert(s: &EditorState, k: &KeyInput) -> EditorState {
    match k.key.as_str() {
        "Escape" => insert::exit_insert(s),
        "Backspace" => insert::backspace(s),
        "Enter" => insert::insert_newline(s),
        "Tab" => insert::insert_tab(s),
        "ArrowLeft" => motions::move_left(s),
        "ArrowRight" => motions::move_right(s),
        "ArrowUp" => motions::move_up(s),
        "ArrowDown" => motions::move_down(s),
        _ => {
            if k.ctrl && (k.key == "w" || k.key == "W") {
                insert::delete_word_before(s)
            } else if k.key.chars().count() == 1 && !k.ctrl && !k.meta {
                insert::insert_text(s, &k.key)
            } else {
                s.clone()
            }
        }
    }
}

fn with_pending(s: &EditorState, p: Pending) -> EditorState {
    let mut ns = s.clone();
    ns.pending = p;
    ns
}

fn handle_normal(s: &EditorState, k: &KeyInput) -> KeyResult {
    match s.pending {
        Pending::G => {
            let s2 = with_pending(s, Pending::None);
            return match k.key.as_str() {
                "g" => state_only(motions::first_line(&s2)),
                "t" => state_effect(s2, AppEffect::TabNext),
                "T" => state_effect(s2, AppEffect::TabPrev),
                _ => handle_normal(&s2, k),
            };
        }
        Pending::D => {
            let s2 = with_pending(s, Pending::None);
            return match k.key.as_str() {
                "d" => state_only(edits::delete_line(&s2)),
                _ => handle_normal(&s2, k),
            };
        }
        Pending::Y => {
            let s2 = with_pending(s, Pending::None);
            return match k.key.as_str() {
                "y" => state_only(edits::yank_line(&s2)),
                _ => handle_normal(&s2, k),
            };
        }
        Pending::None => {}
    }

    // Navigation effects (Phase 2): reachable in NORMAL mode only.
    match k.key.as_str() {
        "[" => return state_effect(s.clone(), AppEffect::PrevDay),
        "]" => return state_effect(s.clone(), AppEffect::NextDay),
        _ => {}
    }
    if k.ctrl && (k.key == "t" || k.key == "T") {
        return state_effect(s.clone(), AppEffect::Today);
    }

    let st = match k.key.as_str() {
        "h" | "ArrowLeft" => motions::move_left(s),
        "l" | "ArrowRight" => motions::move_right(s),
        "j" | "ArrowDown" => motions::move_down(s),
        "k" | "ArrowUp" => motions::move_up(s),
        "w" => motions::word_forward(s),
        "b" => motions::word_backward(s),
        "e" => motions::word_end(s),
        "0" => motions::line_start(s),
        "$" => motions::line_end(s),
        "G" => motions::last_line(s),
        "g" => with_pending(s, Pending::G),
        "d" => with_pending(s, Pending::D),
        "y" => with_pending(s, Pending::Y),
        "x" => edits::delete_char(s),
        "p" => edits::paste_below(s),
        "P" => edits::paste_above(s),
        "t" => edits::toggle_todo(s),
        "u" => state::undo(s),
        "i" => insert::enter_insert(s),
        "a" => insert::enter_insert_after(s),
        "A" => insert::enter_insert_line_end(s),
        "o" => insert::open_below(s),
        "O" => insert::open_above(s),
        "Enter" => motions::move_down(s),
        // ":" command mode is deferred to Phase 5.
        _ => {
            if k.ctrl && (k.key == "r" || k.key == "R") {
                state::redo(s)
            } else {
                s.clone()
            }
        }
    };
    state_only(st)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::editor::state::{Cursor, create_editor_state};

    fn key(k: &str) -> KeyInput {
        KeyInput {
            key: k.into(),
            ctrl: false,
            meta: false,
            shift: false,
        }
    }

    #[test]
    fn i_enters_insert_then_typing_inserts() {
        let s = create_editor_state(vec!["ac".into()], vec![]);
        let s = handle_key(&s, &key("i")).state;
        assert_eq!(s.mode, Mode::Insert);
        let s = handle_key(&s, &key("x")).state;
        assert_eq!(s.lines, vec!["xac".to_string()]);
    }

    #[test]
    fn dd_requires_two_keystrokes() {
        let s = create_editor_state(vec!["one".into(), "two".into()], vec![]);
        let s = handle_key(&s, &key("d")).state;
        assert_eq!(s.pending, Pending::D);
        let s = handle_key(&s, &key("d")).state;
        assert_eq!(s.lines, vec!["two".to_string()]);
    }

    #[test]
    fn ctrl_r_redoes() {
        let mut s = create_editor_state(vec!["a".into()], vec![]);
        s = handle_key(&s, &key("x")).state; // delete 'a' (pushes undo)
        s = handle_key(&s, &key("u")).state; // undo
        assert_eq!(s.lines, vec!["a".to_string()]);
        let r = handle_key(
            &s,
            &KeyInput {
                key: "r".into(),
                ctrl: true,
                meta: false,
                shift: false,
            },
        )
        .state;
        assert_eq!(r.lines, vec![String::new()]);
    }

    // Ported from web/src/lib/editor/keymap.test.ts (editing subset only)

    #[test]
    fn j_moves_down_in_normal_mode() {
        let s = create_editor_state(vec!["a".into(), "b".into()], vec![]);
        let r = handle_key(&s, &key("j")).state;
        assert_eq!(r.cursor.line, 1);
    }

    #[test]
    fn escape_exits_insert_mode() {
        let s = create_editor_state(vec!["".into()], vec![]);
        let s = handle_key(&s, &key("i")).state;
        assert_eq!(s.mode, Mode::Insert);
        let s = handle_key(&s, &key("x")).state;
        assert_eq!(s.lines, vec!["x".to_string()]);
        let s = handle_key(&s, &key("Escape")).state;
        assert_eq!(s.mode, Mode::Normal);
    }

    #[test]
    fn gg_jumps_to_first_line() {
        let mut s = create_editor_state(vec!["a".into(), "b".into(), "c".into()], vec![]);
        s.cursor = Cursor { line: 2, col: 0 };
        let s = handle_key(&s, &key("g")).state;
        let s = handle_key(&s, &key("g")).state;
        assert_eq!(s.cursor.line, 0);
    }

    // Ported from web/src/lib/editor/keymap.test.ts — navigation effects.

    #[test]
    fn gt_emits_tab_next_effect() {
        let s = create_editor_state(vec!["a".into()], vec![]);
        let s = handle_key(&s, &key("g")).state;
        assert_eq!(handle_key(&s, &key("t")).effect, Some(AppEffect::TabNext));
    }

    #[test]
    fn shift_gt_emits_tab_prev_effect() {
        let s = create_editor_state(vec!["a".into()], vec![]);
        let s = handle_key(&s, &key("g")).state;
        assert_eq!(handle_key(&s, &key("T")).effect, Some(AppEffect::TabPrev));
    }

    #[test]
    fn bracket_keys_emit_day_navigation() {
        let s = create_editor_state(vec!["a".into()], vec![]);
        assert_eq!(handle_key(&s, &key("[")).effect, Some(AppEffect::PrevDay));
        assert_eq!(handle_key(&s, &key("]")).effect, Some(AppEffect::NextDay));
    }

    #[test]
    fn ctrl_t_emits_today_effect() {
        let s = create_editor_state(vec!["a".into()], vec![]);
        let r = handle_key(
            &s,
            &KeyInput {
                key: "t".into(),
                ctrl: true,
                meta: false,
                shift: false,
            },
        );
        assert_eq!(r.effect, Some(AppEffect::Today));
    }

    // TODO(phase 5): command mode (`:`) tests — deferred to Phase 5

    // Review flag: keymap.ts:137-138 maps Escape in NORMAL mode to enterInsert(...).
    // This looks like a latent bug and is deliberately NOT replicated here.
}
